test/utilities.jl

```diff
@@ -0,0 +1,128 @@
+"""
+assert_allclose(actual, reference; [rtol], [atol], [err_msg], [error_location_fmt])
+
+Raises an assertion exception when any entries in `actual` and `reference` are not equal up to the
+desired tolerance. In other words, all entries of `actual` and `reference` must satisfy:
+`abs(actual - reference) <= rtol * abs(reference) + atol`
+
+This is inspired by numpy's assert_allclose function. However, we've changed some default values.
+Additionally, in cases where actual and reference don't have equal values but do have the same 
+shape, the locations of the max errors are also printed.
+
+# Keywords
+- `rtol`: The desired relative tolerance
+- `atol`: The desired absolute tolerance
+- `err_msg`: An optional error message that can be printed when the arrays are not close
+- `error_location_fmt`: An optional keyword used to specify a function that creates a string 
+  describing the location in the arrays where values are unequal, given a `CartesianIndex` object.
+  This should not be specified in cases where actual and reference have different shapes.
+"""
+function assert_allclose(actual, reference; rtol = 1e-7, atol = 0.0, err_msg = nothing,
+                         error_location_fmt = nothing)
+    # make the initial check fast!
+    if !any( abs.(actual .- reference) .> (rtol .* abs.(reference) .+ atol))
+        return true
+    end
+
+    common_layout = (size(actual) == size(reference)) && (ndims(actual) == ndims(reference))
```

Comment:
Isn't the second part redundant?

Reply:
You're right. I was thinking of numpy's `size` and `ndim` functions when I wrote that